### PR TITLE
support static executable

### PR DESCRIPTION
**Pull request by miho**
Hi, I added a STATIC_BUILD option to simplify distribution of the appx binary.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -13,6 +13,16 @@ enable_testing()
 project(appx)
 
 include(CheckCXXSourceCompiles)
+
+option(STATIC_BUILD "creates a static executable" OFF)
+
+if(STATIC_BUILD)
+ #MESSAGE("STATIC_BUILD=ON: creating static executable")
```
Comment (smeenai) [inline CMakeLists.txt:20]: This should be deleted.

File: CMakeLists.txt [hunk 2]
```diff
@@ -29,12 +39,24 @@ target_include_directories(appx
                            PrivateHeaders
                            ${OPENSSL_INCLUDE_DIR}
                            ${ZLIB_INCLUDE_DIRS})
-target_link_libraries(appx
+						   
+						   
+if(STATIC_BUILD)
+ target_link_libraries(appx
```
Comment (smeenai) [inline CMakeLists.txt:45]: Style nits: the rest of the file uses a two space indent and aligns continuation lines one column after the opening parenthesis.

File: CMakeLists.txt [hunk 3]
```diff
@@ -13,6 +13,16 @@ enable_testing()
 project(appx)
 
 include(CheckCXXSourceCompiles)
+
+option(STATIC_BUILD "creates a static executable" OFF)
+
+if(STATIC_BUILD)
+ #MESSAGE("STATIC_BUILD=ON: creating static executable")
+ SET(CMAKE_FIND_LIBRARY_SUFFIXES ".a")
```
Comment (smeenai) [inline CMakeLists.txt:21]: Style nit: cmake commands should be lowercase.

File: CMakeLists.txt [hunk 4]
```diff
@@ -29,12 +39,24 @@ target_include_directories(appx
...
                            ${OPENSSL_INCLUDE_DIR}
                            ${ZLIB_INCLUDE_DIRS})
-target_link_libraries(appx
+						   
+						   
+if(STATIC_BUILD)
+ target_link_libraries(appx
+                      PRIVATE
+                      ${OPENSSL_LIBRARIES}
+                      ${ZLIB_LIBRARIES} -ldl -static-libgcc -static-libstdc++)
```
Comment (smeenai) [inline CMakeLists.txt:48]: Do you need the `-static-libgcc -static-libstdc++` explicitly? Doesn't passing `-static` achieve that automatically?

File: CMakeLists.txt [hunk 5]
```diff
@@ -29,12 +39,24 @@ target_include_directories(appx
...
+else()
+ target_link_libraries(appx
                       PRIVATE
                       ${OPENSSL_LIBRARIES}
-                      ${ZLIB_LIBRARIES})
+                      ${ZLIB_LIBRARIES} -ldl )
```
Comment (smeenai) [inline CMakeLists.txt:53]: The `-ldl` addition seems unrelated to static builds.

File: CMakeLists.txt [hunk 6]
```diff
@@ -108,3 +130,18 @@ appx_add_test(TestXMLEscaping)
 appx_add_test(TestZIPEscaping)
 appx_add_test(TestContentTypes)
 appx_add_test(TestEmptyFile)
+
+MESSAGE(STATUS "--------------------------------------------------------------------------------")
```
Comment (smeenai) [inline CMakeLists.txt:134]: I'm not a big fan of status outputs like this during configuration. They're helpful for debugging but too much information to have every single configure.

File: README.md [hunk 7]
```diff
@@ -28,12 +28,47 @@ Build:
...
     mkdir Build && cd Build && cmake .. && make
 
-On OS X, you'll need to explicitly point cmake to your OpenSSL
+On macOS, you'll need to explicitly point cmake to your OpenSSL
 installation. The easiest method is to install OpenSSL using
 [Homebrew](http://brew.sh/) and then pass `-DOPENSSL_ROOT_DIR=$(brew --prefix openssl)`
 when invoking cmake. You can also compile OpenSSL [from source](https://github.com/openssl/openssl)
 and set `OPENSSL_ROOT_DIR` accordingly.
 
+The `-DSTATIC_BUILD=ON` option is not supported on macOS. To create a static (statically links zlib and openssl) `appx` executable on macOS the following cmake command can be used:
```
Comment (smeenai) [inline README.md:37]: There should probably be a description of `-DSTATIC_BUILD=ON` before this.

**Review by smeenai — COMMENTED** [review]
Looks good for the most part. Thanks for working on this!

**Comment by strager** [comment]
Why is this part of the fb-util-for-appx project?

From my understanding, building a static executable is the responsibility of whoever's compiling fb-util-for-appx. (Of course, fb-util-for-appx should try to not get in the way. I'd definitely accept changes which fix assumptions in fb-util-for-appx!)

Ideally you'd just run the following to build a static executable on Linux:

    $ mkdir build
    $ cd build
    $ cmake -DCMAKE_EXE_LINKER_FLAGS=-static ..
    $ make

When I run this on my machine, I hit a problem with OpenSSL (and adding -ldl (as you've done) fixes it). I think we just need to tackle this issue to make static builds work on Linux. I'll do some investigation.

**Comment by strager** [comment]
On my Linux system (CentOS), `pkg-config --libs --static openssl` gives me flags which cannot work with `-static`. (openssl depends on krb5 which is only a DSO.) So even if FindOpenSSL used pkg-config properly, it still wouldn't work for your use case.

CMake 3.7.2's FindOpenSSL [1] is inflexible and won't let the user override OPENSSL_LIBRARIES. You can override OPENSSL_SSL_LIBRARY and OPENSSL_CRYPTO_LIBRARY, but you can only override them with a single path each (and not `-lcrypto -ldl -lz`, for example).

@miho, do you know if any other projects which use OpenSSL support `-static`? I'm curious how they solve or work around these problems.

[1] My Linux system has CMake 3.7.2 installed. I didn't check other versions of CMake.

**Comment by miho** [comment]
@strager 

> Why is this part of the fb-util-for-appx project?

I added it since almost every non-trivial project needs some extra attention for static compilation. But for me it actually turned out to be sufficient to explicitly specify *.a libs as explained in the README.

> do you know if any other projects which use OpenSSL support -static?

unfortunately no. I didn't expect any problems since static versions of SSL and CRYPTO seem to work on Debian/Ubuntu and Homebrew on OSX.

**Comment by strager** [comment]
I managed to hackily build with -static (without any platform-specific changes to fb-util-for-appx):

```
$ mkdir static-openssl
$ printf 'GROUP(/usr/lib64/libssl.a -lcrypto -ldl -lz)\n' >static-openssl/libssl.a
$ printf 'GROUP(/usr/lib64/libcrypto.a -ldl -lz)\n' >static-openssl/libcrypto.a
$ mkdir build
$ cd build/
$ OPENSSL_ROOT_DIR="${PWD}/../static-openssl" cmake -DOPENSSL_USE_STATIC_LIBS=ON -DOPENSSL_STATIC_FOUND=OFF -DCMAKE_EXE_LINKER_FLAGS=-static ..
$ make
```

We make some linker scripts which link OpenSSL properly (platform-specific). Then we make FindOpenSSL find our linker scripts instead of the OpenSSL libraries installed on the system.

@miho, would something like the above work for you?

I'm concerned with adding `-ldl` to fb-util-for-appx when `-ldl` is a requirement specific to *your* configuration of OpenSSL. pkg-config is meant to solve this, but pkg-config is insufficient on my system.

Note that using `-static` with `-ldl` is problematic. On my system, the linker prints a warning:

    (.text+0x4b9): warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking

It sounds like this goes against the original intent of your pull request: "to simplify distribution of the appx binary".

**Comment by miho** [comment]
> would something like the above work for you?

Yes that would be a sufficient solution. 

> It sounds like this goes against the original intent of your pull request: "to simplify distribution of the appx binary".

I agree.  I just tried to run the compiled binary on different versions of Ubuntu (including Microsofts new Ubuntu bash integration). Binaries with `-static` didn't run on all of them. With just static SSL/CRYPTO and ZLIB it worked fine. So that's probably the best solution.

**Comment by strager** [comment]
> unfortunately no. I didn't expect any problems since static versions of SSL and CRYPTO seem to work on Debian/Ubuntu and Homebrew on OSX.

If it worked without any problems, why did you add `-ldl -static-libgcc -static-libstdc++` in commit e947834103f05aa78a46feda5442761e8f6c0aea and `-ldl` in commit f1d724f0cc0f1f8b12ec71ba189331087febb81f?

>> would something like the above work for you?
>
> Yes that would be a sufficient solution.

Okay. If we want to support `-static`, it sounds like we just need documentation explaining how to build with `-static` (and no build system changes).

> With just static SSL/CRYPTO and ZLIB it worked fine. So that's probably the best solution.

Alright. I'll check tomorrow if just `-DOPENSSL_USE_STATIC_LIBS=ON` is necessary or if more steps are needed.

**Comment by strager** [comment]
I tried to make CMake wrap the output of `pkg-config --libs --static openssl` in `-Wl,-Bstatic ... -Wl,-Bdynamic` automatically. `LINK_SEARCH_START_STATIC` is the tool for this I think, but it doesn't work with `INTERFACE` or `IMPORTED` targets.

I then hacked in the flags manually, and that didn't work because I don't have `libkrb5.a`. (I assumed `-Bstatic`/`-Bdynamic` change the *default* search extensions. I was wrong.)

Even if I manually implement (in CMake) the logic I was looking for, we end up trying to link libdl.a, which is a no-no.

I don't have any other ideas on automatically linking OpenSSL as-statically-as-possible using CMake, pkg-config, or GNU ld/gold (short of putting platform-specific hacks into fb-util-for-appx). So I think we need to leave it up to the person building the project.

**Comment by strager** [comment]
Okay, I have something pretty hacky (yet probably portable) working. I'll do some more testing and upload it.

**Comment by strager** [comment]
I think this patch is very distro-specific, so it doesn't belong in this project's repository. Sorry!
